File: java/core/src/main/java/com/github/pukkaone/mullet/html/ElementRenderer.java
/*
Copyright (c) 2011, Chin Huang
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

  * Redistributions of source code must retain the above copyright notice,
    this list of conditions and the following disclaimer.
  * Redistributions in binary form must reproduce the above copyright notice,
    this list of conditions and the following disclaimer in the documentation
    and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.github.pukkaone.mullet.html;

import com.github.pukkaone.mullet.Container;
import com.github.pukkaone.mullet.RenderContext;
import com.github.pukkaone.mullet.Renderer;
import com.github.pukkaone.mullet.TemplateException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.StringTokenizer;

/**
 * Renders an HTML element.
 */
class ElementRenderer extends Container implements Renderer {

    private static final String ATTRIBUTE_SEPARATOR = ";";
    private static final char ATTRIBUTE_NAME_SEPARATOR = '=';
    private static final String ATTRIBUTE_SYNTAX_ERROR =
            "expected '%s' in '%s'";
    private static final String ATTRIBUTE_NAME_MISSING_ERROR =
            "attribute name missing in '%s'";
    private static final String[] CONVENIENT_ATTRIBUTE_COMMANDS =
            new String[] {
                Command.ACTION,
                Command.ALT,
                Command.HREF,
                Command.SRC,
                Command.TITLE,
                Command.VALUE,
            };

    private Element element;
    private ArrayList<AttributeCommand> attributeCommands =
            new ArrayList<AttributeCommand>();
    RemoveMode removeMode;
    private String textVariableName;
    private Message textMessage;
    private Template template;
    private Boolean escapeXml;

    ElementRenderer(Element element) {
        this.element = element;
    }

    private void addAttributeCommand(
            String attributeName, String variableName)
    {
        attributeCommands.add(
                new ModelAttributeCommand(attributeName, variableName));
    }

    private void addAttributeCommands(String attributeAndVariablePairs) {
        StringTokenizer iAttribute = new StringTokenizer(
                attributeAndVariablePairs, ATTRIBUTE_SEPARATOR);
        while (iAttribute.hasMoreTokens()) {
            String commandText = iAttribute.nextToken();

            int iColon = commandText.indexOf(ATTRIBUTE_NAME_SEPARATOR);
            if (iColon < 0) {
                throw new TemplateException(String.format(
                        ATTRIBUTE_SYNTAX_ERROR,
                        ATTRIBUTE_NAME_SEPARATOR,
                        commandText));
            }

            String attributeName = commandText.substring(0, iColon).trim();
            if (attributeName.length() == 0) {
                throw new TemplateException(String.format(
                        ATTRIBUTE_NAME_MISSING_ERROR, commandText));
            }

            String variableName = commandText.substring(iColon + 1).trim();
            if (variableName.length() == 0) {
                throw new TemplateException(String.format(
                        "variable name missing in '%s'", commandText));
            }

            addAttributeCommand(attributeName, variableName);
        }
    }

    private void configureAttributeCommands(Attributes commandAttributes) {
        String value = commandAttributes.get(Command.ATTR);
        if (value != null) {
            addAttributeCommands(value);
        }

        for (String attributeName : CONVENIENT_ATTRIBUTE_COMMANDS) {
            value = commandAttributes.get(attributeName);
            if (value != null) {
                addAttributeCommand(attributeName, value);
            }
        }

        value = commandAttributes.get(Command.REMOVE);
        if (value != null) {
            value = value.trim();
            try {
                removeMode = RemoveMode.valueOf(value.toUpperCase());
            } catch (IllegalArgumentException e) {
                throw new TemplateException(
                        String.format("invalid remove argument '%s'", value),
                        e);
            }
        }
    }

    private void addAttributeMessageCommand(String attributeName,
            String messageArguments) {
        attributeCommands.add(
                new MessageAttributeCommand(
                        attributeName, new Message(messageArguments)));
    }

    private void addAttributeMessageCommands(String attributeAndMessagePairs) {
        StringTokenizer iAttribute = new StringTokenizer(
                attributeAndMessagePairs, ATTRIBUTE_SEPARATOR);
        while (iAttribute.hasMoreTokens()) {
            String commandText = iAttribute.nextToken();

            int iColon = commandText.indexOf(ATTRIBUTE_NAME_SEPARATOR);
            if (iColon < 0) {
                throw new TemplateException(String.format(
                        ATTRIBUTE_SYNTAX_ERROR,
                        ATTRIBUTE_NAME_SEPARATOR,
                        commandText));
            }

            String attributeName = commandText.substring(0, iColon).trim();
            if (attributeName.length() == 0) {
                throw new TemplateException(String.format(
                        ATTRIBUTE_NAME_MISSING_ERROR, commandText));
            }

            String messageArguments = commandText.substring(iColon + 1).trim();
            if (messageArguments.length() == 0) {
                throw new TemplateException(String.format(
                        "message arguments missing in '%s'", commandText));
            }

            addAttributeMessageCommand(attributeName, messageArguments);
        }
    }

    private void configureAttributeMessageCommands(Attributes commandAttributes) {
        String value = commandAttributes.get(Command.ALT_MESSAGE);
        if (value != null) {
            addAttributeMessageCommand(Command.ALT, value);
        }

        value = commandAttributes.get(Command.ATTR_MESSAGE);
        if (value != null) {
            addAttributeMessageCommands(value);
        }

        value = commandAttributes.get(Command.TITLE_MESSAGE);
        if (value != null) {
            addAttributeMessageCommand(Command.TITLE, value);
        }

        value = commandAttributes.get(Command.VALUE_MESSAGE);
        if (value != null) {
            addAttributeMessageCommand(Command.VALUE, value);
        }
    }

    private void configureContent(
            Attributes commandAttributes, TemplateLoader loader)
    {
        textVariableName = commandAttributes.get(Command.TEXT);
        if (textVariableName != null) {
            textVariableName = textVariableName.intern();
            return;
        }

        String textMessageArguments =
                commandAttributes.get(Command.TEXT_MESSAGE);
        if (textMessageArguments != null) {
            textMessage = new Message(textMessageArguments);
            return;
        }

        String uri = commandAttributes.get(Command.INCLUDE);
        if (uri != null) {
            try {
                template = loader.load(uri);
            } catch (IOException e) {
                throw new TemplateException("load", e);
            }
        }
    }

    void configureCommands(
            Attributes commandAttributes, TemplateLoader loader)
    {
        configureAttributeCommands(commandAttributes);
        configureAttributeMessageCommands(commandAttributes);

        configureContent(commandAttributes, loader);

        String escapeXmlValue = commandAttributes.get(Command.ESCAPE_XML);
        if (escapeXmlValue != null) {
            escapeXml = !Boolean.FALSE.toString().equals(escapeXmlValue);
        }
    }

    /**
     * Checks if this renderer has a command.  If it has a command, then the
     * template builder should not discard it.
     */
    boolean hasCommand() {
        return !attributeCommands.isEmpty();
    }

    /**
     * Checks if the element content will be rendered by a command.
     */
    boolean hasDynamicContent() {
        return textVariableName != null
            || textMessage != null
            || template != null;
    }

    private Attributes executeAttributeCommands(RenderContext renderContext) {
        if (attributeCommands.isEmpty()) {
            return element.attributes;
        }

        // Copy the original attributes.  The commands modify the copy to
        // produce the attributes to render.
        Attributes renderAttributes = (Attributes) element.attributes.clone();
        for (AttributeCommand command : attributeCommands) {
            command.execute(renderContext, renderAttributes);
        }
        return renderAttributes;
    }

    private boolean shouldRenderTag() {
        return removeMode == null || removeMode == RemoveMode.CONTENT;
    }

    private void renderStartTag(RenderContext renderContext) {
        if (shouldRenderTag()) {
            Attributes attributes = executeAttributeCommands(renderContext);
            renderContext.write(element.renderStartTag(attributes));
        }
    }

    private void renderEndTag(RenderContext renderContext) {
        if (shouldRenderTag()) {
            renderContext.write(element.renderEndTag());
        }
    }

    private boolean shouldRenderContent() {
        return removeMode != RemoveMode.CONTENT;
    }

    private void renderContent(RenderContext renderContext) {
        if (shouldRenderContent()) {
            if (textVariableName != null) {
                Object value = renderContext.getDisplayValue(textVariableName);
                String text = renderContext.escapeXml(value.toString());
                renderContext.write(text);
            } else if (textMessage != null) {
                String text = textMessage.format(renderContext);
                text = renderContext.escapeXml(text);
                renderContext.write(text);
            } else if (template != null) {
                template.render(renderContext);
            } else {
                renderChildren(renderContext);
            }
        }
    }

    public void render(RenderContext renderContext) {
        // Process the command to change the escaping mode.
        boolean originalEscapeXmlEnabled = renderContext.isEscapeXmlEnabled();
        if (escapeXml != null) {
            renderContext.setEscapeXmlEnabled(escapeXml);
        }

        renderStartTag(renderContext);
        renderContent(renderContext);
        renderEndTag(renderContext);

        // Restore the original escaping mode.
        if (escapeXml != null) {
            renderContext.setEscapeXmlEnabled(originalEscapeXmlEnabled);
        }
    }
}

File: ruby/core/lib/mullet/html/element_renderer.rb
require 'mullet/container'
require 'mullet/template_error'
require 'mullet/html/command'
require 'mullet/html/message'
require 'mullet/html/message_attribute_command'
require 'mullet/html/model_attribute_command'

module Mullet; module HTML

  # Renders an HTML element.
  class ElementRenderer
    include Container

    ATTRIBUTE_SEPARATOR = ';'
    ATTRIBUTE_NAME_SEPARATOR = '='
    ATTRIBUTE_SYNTAX_ERROR = "expected '%s' in '%s'"
    ATTRIBUTE_NAME_MISSING_ERROR = "attribute name missing in '%s'"
    CONVENIENT_ATTRIBUTE_COMMANDS = [
        Command::ACTION,
        Command::ALT,
        Command::HREF,
        Command::SRC,
        Command::TITLE,
        Command::VALUE ]

    attr_reader :remove_mode

    # Constructor
    #
    # @param [Element] element
    #           element to render
    def initialize(element)
      super()

      @element = element
      @attribute_commands = []
      @remove_mode = nil
      @text_variable_name = nil
      @text_message = nil
      @template = nil
      @escape_xml = nil
    end

    def add_attribute_command(attribute_name, variable_name)
      @attribute_commands <<
          ModelAttributeCommand.new(attribute_name, variable_name)
    end

    def add_attribute_commands(attribute_variable_pairs)
      attribute_variable_pairs.split(ATTRIBUTE_SEPARATOR).each do |command_text|
        command_parts = command_text.split(ATTRIBUTE_NAME_SEPARATOR, 2)
        if command_parts.length() < 2
          raise TemplateError.new(
              ATTRIBUTE_SYNTAX_ERROR % [ATTRIBUTE_NAME_SEPARATOR, command_text])
        end

        attribute_name = command_parts[0].strip()
        if attribute_name.empty?()
          raise TemplateError.new(ATTRIBUTE_NAME_MISSING_ERROR % command_text)
        end

        variable_name = command_parts[1].strip()
        if variable_name.empty?()
          raise TemplateError.new("variable name missing in '#{command_text}'")
        end

        add_attribute_command(attribute_name, variable_name)
      end
    end

    def configure_attribute_commands(command_attributes)
      value = command_attributes.fetch(Command::ATTR, nil)
      if value != nil
        add_attribute_commands(value)
      end

      CONVENIENT_ATTRIBUTE_COMMANDS.each do |attribute_name|
        value = command_attributes.fetch(attribute_name, nil)
        if value != nil
          add_attribute_command(attribute_name, value)
        end
      end

      value = command_attributes.fetch(Command::REMOVE, nil)
      if value != nil
        value.strip!();
        value.downcase!()
        @remove_mode = RemoveMode.value_of(value)
        if @remove_mode == nil
          raise TemplateError.new("invalid remove argument '#{value}'")
        end
      end
    end

    def add_attribute_message_command(attribute_name, message_arguments)
      @attribute_commands << MessageAttributeCommand.new(
          attribute_name, Message.new(message_arguments))
    end

    def add_attribute_message_commands(attribute_message_pairs)
      attribute_message_pairs.split(ATTRIBUTE_SEPARATOR).each do |command_text|
        command_parts = command_text.split(ATTRIBUTE_NAME_SEPARATOR, 2)
        if parts.length() < 2
          raise TemplateError.new(
              ATTRIBUTE_SYNTAX_ERROR % [ATTRIBUTE_NAME_SEPARATOR, command_text])
        end

        attribute_name = command_parts[0].strip()
        if attribute_name.empty?()
          raise TemplateError.new(ATTRIBUTE_NAME_MISSING_ERROR % command_text)
        end

        message_arguments = command_parts[1].strip()
        if message_arguments.empty?()
          raise TemplateError.new(
              "message arguments missing in '#{command_text}'")
        end

        add_attribute_message_command(attribute_name, message_arguments)
      end
    end

    def configure_attribute_message_commands(command_attributes)
      value = command_attributes.fetch(Command::ALT_MESSAGE, nil)
      if value != nil
        add_attribute_message_command(Command::ALT, value)
      end

      value = command_attributes.fetch(Command::ATTR_MESSAGE, nil)
      if value != nil
        add_attribute_message_commands(value)
      end

      value = command_attributes.fetch(Command::TITLE_MESSAGE, nil)
      if value != nil
        add_attribute_message_command(Command::TITLE, value)
      end

      value = command_attributes.fetch(Command::VALUE_MESSAGE, nil)
      if value != nil
        add_attribute_message_command(Command::VALUE, value)
      end
    end

    def configure_content(command_attributes, template_loader)
      @text_variable_name = command_attributes.fetch(Command::TEXT, nil)
      if @text_variable_name != nil
        @text_variable_name = @text_variable_name.to_sym()
        return
      end

      text_message_arguments =
          command_attributes.fetch(Command::TEXT_MESSAGE, nil)
      if text_message_arguments != nil
        @text_message = Message.new(text_message_arguments)
        return
      end

      uri = command_attributes.fetch(Command::INCLUDE, nil)
      if uri != nil
        @template = template_loader.load(uri)
      end
    end

    def configure_commands(command_attributes, template_loader)
      configure_attribute_commands(command_attributes)
      configure_attribute_message_commands(command_attributes)

      configure_content(command_attributes, template_loader)

      escape_xml_value = command_attributes.fetch(Command::ESCAPE_XML, nil)
      if escape_xml_value != nil
        @escape_xml = escape_xml_value != 'false' 
      end
    end

    # Checks if this renderer has a command.  If it has a command, then the
    # template builder should not discard it.
    def has_command()
      return !@attribute_commands.empty?()
    end

    # Checks if the element content will be rendered by a command.
    def has_dynamic_content()
      return @text_variable_name != nil ||
             @text_message != nil ||
             @template != nil
    end

    def execute_attribute_commands(render_context)
      if @attribute_commands.empty?()
        return @element.attributes()
      end

      # Copy the original attributes.  The commands modify the copy to
      # produce the attributes to render.
      render_attributes = @element.attributes().dup()
      @attribute_commands.each do |command|
        command.execute(render_context, render_attributes)
      end
      return render_attributes
    end

    def should_render_tag()
      return @remove_mode == nil || @remove_mode == RemoveMode::CONTENT
    end

    def render_start_tag(render_context)
      if should_render_tag()
        attributes = execute_attribute_commands(render_context)
        render_context << @element.render_start_tag(attributes)
      end
    end

    def render_end_tag(render_context)
      if should_render_tag()
        render_context << @element.render_end_tag()
      end
    end

    def should_render_content()
      return @remove_mode != RemoveMode::CONTENT
    end

    def render_content(render_context)
      if should_render_content()
        if @text_variable_name != nil
          value = render_context.get_display_value(@text_variable_name)
          text = render_context.escape_xml(value.to_s())
          render_context << text
        elsif @text_message != nil
          text = @text_message.format(render_context)
          text = render_context.escape_xml(text)
          render_context << text
        elsif @template != nil
          @template.render(render_context)
        else
          render_children(render_context)
        end
      end
    end

    def render(render_context)
      # Process the command to change the escaping mode.
      original_escape_xml_enabled = render_context.escape_xml_enabled()
      if @escape_xml != nil
        render_context.escape_xml_enabled = @escape_xml
      end

      render_start_tag(render_context)
      render_content(render_context)
      render_end_tag(render_context)

      # Restore the original escaping mode.
      if @escape_xml != nil
        render_context.escape_xml_enabled = original_escape_xml_enabled
      end
    end
  end

end; end
